File: examples/policy-engine-service-wasm-example.ts
/**
 * Example demonstrating how to use the Policy Engine service with OPA WASM integration in TypeScript
 *
 * This example shows how to:
 * 1. Create a policy with GDPR rules
 * 2. Evaluate data against the policy using the service API
 * 3. Understand policy decisions and reasons
 * 4. Update and manage policies
 *
 * Note: If the Policy Engine service is not running, this example will automatically
 * switch to mock mode and simulate the API responses.
 */

import axios from 'axios';
import { v4 as uuidv4 } from 'uuid';
import fs from 'fs';
import path from 'path';

// Define the base URL for the Policy Engine service
const BASE_URL = 'http://localhost:8001';

// Flag to track if we're in mock mode
let MOCK_MODE = false;

// Define interfaces for the example
interface Policy {
  name: string;
  description: string;
  version: string;
  type: string;
  rules: any[];
  metadata?: Record<string, any>;
}

interface PolicyResponse {
  policy_id: string;
  policy: Policy;
}

interface EvaluationRequest {
  input_data: Record<string, any>;
}

interface EvaluationResponse {
  allowed: boolean;
  reasons: string[];
  errors?: string[];
}

interface TestCase {
  name: string;
  data: Record<string, any>;
  expected_result: boolean;
}

// Helper function to check if the service is running
async function checkServiceAvailability(): Promise<boolean> {
  try {
    await axios.get(`${BASE_URL}/health`, { timeout: 2000 });
    return true;
  } catch (error: any) {
    console.log(
      'Policy Engine service is not available. Switching to mock mode.'
    );
    MOCK_MODE = true;
    return false;
  }
}

// Helper function to create a policy
async function createPolicy(policy: Policy): Promise<PolicyResponse | null> {
  if (MOCK_MODE) {
    return mockCreatePolicy(policy);
  }

  try {
    const response = await axios.post(`${BASE_URL}/policies`, policy);
    return response.data;
  } catch (error: any) {
    console.error('Error creating policy:', error.message);
    console.log('Switching to mock mode.');
    MOCK_MODE = true;
    return mockCreatePolicy(policy);
  }
}

// Mock implementation of createPolicy
function mockCreatePolicy(policy: Policy): PolicyResponse {
  const policyId = uuidv4();
  console.log('MOCK: Created policy with ID:', policyId);

  return {
    policy_id: policyId,
    policy: policy,
  };
}

// Helper function to evaluate a policy
async function evaluatePolicy(
  policyId: string,
  data: Record<string, any>
): Promise<EvaluationResponse | null> {
  if (MOCK_MODE) {
    return mockEvaluatePolicy(data);
  }

  try {
    const response = await axios.post(
      `${BASE_URL}/policies/${policyId}/evaluate`,
      { input_data: data }
    );
    return response.data;
  } catch (error: any) {
    console.error('Error evaluating policy:', error.message);
    console.log('Switching to mock mode.');
    MOCK_MODE = true;
    return mockEvaluatePolicy(data);
  }
}

// Mock implementation of evaluatePolicy
function mockEvaluatePolicy(data: Record<string, any>): EvaluationResponse {
  // Simulate GDPR policy evaluation
  const user = data.user || {};
  const purpose = data.purpose;
  const context = data.context || {};

  // Check conditions
  const hasConsent = user.consent === true;
  const hasPurpose = purpose !== null && purpose !== undefined;
  const isPurposeAllowed =
    hasPurpose &&
    Array.isArray(user.allowed_purposes) &&
    user.allowed_purposes.includes(purpose);
  const isSensitivityAcceptable =
    !context.sensitivityLevel || context.sensitivityLevel <= 3;

  // Determine if allowed
  const allowed =
    hasConsent && hasPurpose && isPurposeAllowed && isSensitivityAcceptable;

  // Generate reasons
  const reasons: string[] = [];

  if (hasConsent) {
    reasons.push('User has provided consent');
  } else {
    reasons.push('User has not provided consent');
  }

  if (hasPurpose) {
    reasons.push('Purpose is specified');

    if (isPurposeAllowed) {
      reasons.push('Purpose is allowed');
    } else {
      reasons.push('Purpose is not allowed');
    }
  } else {
    reasons.push('Purpose is not specified');
  }

  if (context.sensitivityLevel && context.sensitivityLevel > 3) {
    reasons.push('Sensitivity level exceeds allowed threshold');
  }

  return {
    allowed,
    reasons,
  };
}

// Helper function to update a policy
async function updatePolicy(
  policyId: string,
  policy: Policy
): Promise<Policy | null> {
  if (MOCK_MODE) {
    return mockUpdatePolicy(policy);
  }

  try {
    const response = await axios.put(
      `${BASE_URL}/policies/${policyId}`,
      policy
    );
    return response.data;
  } catch (error: any) {
    console.error('Error updating policy:', error.message);
    console.log('Switching to mock mode.');
    MOCK_MODE = true;
    return mockUpdatePolicy(policy);
  }
}

// Mock implementation of updatePolicy
function mockUpdatePolicy(policy: Policy): Policy {
  console.log('MOCK: Updated policy');
  // Return a deep copy of the policy to ensure all properties are preserved
  return {
    ...policy,
    version: policy.version || '1.0.0', // Ensure version is defined
  };
}

// Helper function to delete a policy
async function deletePolicy(policyId: string): Promise<boolean> {
  if (MOCK_MODE) {
    return mockDeletePolicy(policyId);
  }

  try {
    await axios.delete(`${BASE_URL}/policies/${policyId}`);
    return true;
  } catch (error: any) {
    console.error('Error deleting policy:', error.message);
    console.log('Switching to mock mode.');
    MOCK_MODE = true;
    return mockDeletePolicy(policyId);
  }
}

// Mock implementation of deletePolicy
function mockDeletePolicy(policyId: string): boolean {
  console.log('MOCK: Deleted policy with ID:', policyId);
  return true;
}

// Main function
async function main() {
  console.log('Policy Engine Service WASM Example (TypeScript)');
  console.log('--------------------------------------------');

  // Check if the service is running
  await checkServiceAvailability();

  if (MOCK_MODE) {
    console.log(
      '\n⚠️  Running in MOCK MODE - Policy Engine service is not available'
    );
    console.log(
      'This example will simulate API responses for demonstration purposes.'
    );
    console.log(
      'To run with the actual service, start the Policy Engine service:'
    );
    console.log('  cd ../services/policy-engine');
    console.log('  poetry run uvicorn app.main:app --reload');
    console.log('');
  }

  // 1. Create a GDPR policy
  console.log('\n1. Creating a GDPR policy...');

  // Define a GDPR policy using JSON Logic
  // These rules will be converted to Rego and compiled to WASM by the service
  const gdprPolicy: Policy = {
    name: 'GDPR Data Processing Policy',
    description: 'Controls data processing based on user consent and purpose',
    version: '1.0.0',
    type: 'compliance',
    rules: [
      // Rule 1: User must have provided consent
      {
        '==': [{ var: 'user.consent' }, true],
      },
      // Rule 2: Purpose must be specified
      {
        '!=': [{ var: 'purpose' }, null],
      },
      // Rule 3: Purpose must be in the list of allowed purposes
      {
        in: [{ var: 'purpose' }, { var: 'user.allowed_purposes' }],
      },
    ],
    metadata: {
      category: 'GDPR',
      owner: 'data-protection-officer',
      tags: ['gdpr', 'consent', 'data-processing'],
    },
  };

  const policyResponse = await createPolicy(gdprPolicy);
  const policyId = policyResponse?.policy_id || uuidv4();

  if (policyResponse) {
    console.log('Policy created successfully:');
    console.log(`  ID: ${policyResponse.policy_id}`);
    console.log(`  Name: ${policyResponse.policy.name}`);
    console.log(`  Version: ${policyResponse.policy.version}`);
  }

  // 2. Evaluate the policy against input data
  console.log('\n2. Evaluating policy against different scenarios...');

  // Test cases
  const testCases: TestCase[] = [
    {
      name: 'Valid consent and purpose',
      data: {
        user: {
          consent: true,
          allowed_purposes: ['marketing', 'analytics', 'service-improvement'],
        },
        purpose: 'marketing',
      },
      expected_result: true,
    },
    {
      name: 'No consent',
      data: {
        user: {
          consent: false,
          allowed_purposes: ['marketing', 'analytics'],
        },
        purpose: 'marketing',
      },
      expected_result: false,
    },
    {
      name: 'Missing purpose',
      data: {
        user: {
          consent: true,
          allowed_purposes: ['marketing', 'analytics'],
        },
        purpose: null,
      },
      expected_result: false,
    },
    {
      name: 'Disallowed purpose',
      data: {
        user: {
          consent: true,
          allowed_purposes: ['marketing', 'analytics'],
        },
        purpose: 'profiling',
      },
      expected_result: false,
    },
  ];

  for (const testCase of testCases) {
    console.log(`\nScenario: ${testCase.name}`);

    const result = await evaluatePolicy(policyId, testCase.data);

    if (result) {
      console.log(`Access allowed: ${result.allowed}`);
      console.log(`Reasons: ${result.reasons.join(', ')}`);

      // Verify the result matches the expected outcome
      if (result.allowed === testCase.expected_result) {
        console.log('✅ Result matches expected outcome');
      } else {
        console.log(
          `❌ Result does not match expected outcome (${testCase.expected_result})`
        );
      }
    }
  }

  // 3. Update the policy to add context-based restrictions
  console.log('\n3. Updating the policy to add context-based restrictions...');

  // Add a new rule for sensitivity level restrictions
  const updatedPolicy = { ...gdprPolicy };
  updatedPolicy.rules.push({
    '<=': [{ var: 'context.sensitivityLevel' }, 3],
  });
  updatedPolicy.version = '1.1.0';
  updatedPolicy.description += ' and sensitivity level';

  const updateResult = await updatePolicy(policyId, updatedPolicy);

  if (updateResult) {
    console.log('Policy updated successfully:');
    console.log(`  New version: ${updateResult.version}`);
    console.log(`  Rules count: ${updateResult.rules?.length || 'unknown'}`);
  }

  // 4. Evaluate with context-based restrictions
  console.log(
    '\n4. Evaluating updated policy with context-based restrictions...'
  );

  const contextTestCases: TestCase[] = [
    {
      name: 'Low sensitivity data',
      data: {
        user: {
          consent: true,
          allowed_purposes: ['marketing', 'analytics'],
        },
        purpose: 'marketing',
        context: {
          sensitivityLevel: 2,
        },
      },
      expected_result: true,
    },
    {
      name: 'High sensitivity data',
      data: {
        user: {
          consent: true,
          allowed_purposes: ['marketing', 'analytics'],
        },
        purpose: 'marketing',
        context: {
          sensitivityLevel: 5,
        },
      },
      expected_result: false,
    },
  ];

  for (const testCase of contextTestCases) {
    console.log(`\nScenario: ${testCase.name}`);

    const result = await evaluatePolicy(policyId, testCase.data);

    if (result) {
      console.log(`Access allowed: ${result.allowed}`);
      console.log(`Reasons: ${result.reasons.join(', ')}`);

      // Verify the result matches the expected outcome
      if (result.allowed === testCase.expected_result) {
        console.log('✅ Result matches expected outcome');
      } else {
        console.log(
          `❌ Result does not match expected outcome (${testCase.expected_result})`
        );
      }
    }
  }
}

main();

File: examples/policy-engine-service-wasm-example.py
#!/usr/bin/env python3
"""
Example demonstrating how to use the Policy Engine service with OPA WASM integration

This example shows how to:
1. Create a policy with GDPR rules
2. Evaluate data against the policy using the service API
3. Understand policy decisions and reasons
4. Update and manage policies
"""

import os
import sys
import json
import requests
import uuid
from pathlib import Path

# Define the base URL for the Policy Engine service
BASE_URL = "http://localhost:8001"

def main():
    print("Policy Engine Service WASM Example")
    print("--------------------------------")
    
    # 1. Create a GDPR policy
    print("\n1. Creating a GDPR policy...")
    
    # Define a GDPR policy using JSON Logic
    # These rules will be converted to Rego and compiled to WASM by the service
    policy_request = {
        "name": "GDPR Data Processing Policy",
        "description": "Controls data processing based on user consent and purpose",
        "version": "1.0.0",
        "type": "compliance",
        "rules": [
            # Rule 1: User must have provided consent
            {
                "==": [{"var": "user.consent"}, True]
            },
            # Rule 2: Purpose must be specified
            {
                "!=": [{"var": "purpose"}, None]
            },
            # Rule 3: Purpose must be in the list of allowed purposes
            {
                "in": [{"var": "purpose"}, {"var": "user.allowed_purposes"}]
            }
        ],
        "metadata": {
            "category": "GDPR",
            "owner": "data-protection-officer",
            "tags": ["gdpr", "consent", "data-processing"]
        }
    }
    
    try:
        response = requests.post(
            f"{BASE_URL}/policies",
            json=policy_request
        )
        response.raise_for_status()
        policy = response.json()
        
        print("Policy created successfully:")
        print(f"  ID: {policy['policy_id']}")
        print(f"  Name: {policy['policy']['name']}")
        print(f"  Version: {policy['policy']['version']}")
        
        # Store policy ID for later use
        policy_id = policy["policy_id"]
        
    except requests.RequestException as e:
        print(f"Error creating policy: {e}")
        # For demo purposes, use a mock ID
        policy_id = str(uuid.uuid4())
        print(f"Using mock policy ID: {policy_id}")
    
    # 2. Evaluate the policy against input data
    print("\n2. Evaluating policy against different scenarios...")
    
    # Test cases
    test_cases = [
        {
            "name": "Valid consent and purpose",
            "data": {
                "user": {
                    "consent": True,
                    "allowed_purposes": ["marketing", "analytics", "service-improvement"]
                },
                "purpose": "marketing"
            },
            "expected_result": True
        },
        {
            "name": "No consent",
            "data": {
                "user": {
                    "consent": False,
                    "allowed_purposes": ["marketing", "analytics"]
                },
                "purpose": "marketing"
            },
            "expected_result": False
        },
        {
            "name": "Missing purpose",
            "data": {
                "user": {
                    "consent": True,
                    "allowed_purposes": ["marketing", "analytics"]
                },
                "purpose": None
            },
            "expected_result": False
        },
        {
            "name": "Disallowed purpose",
            "data": {
                "user": {
                    "consent": True,
                    "allowed_purposes": ["marketing", "analytics"]
                },
                "purpose": "profiling"
            },
            "expected_result": False
        }
    ]
    
    for test_case in test_cases:
        print(f"\nScenario: {test_case['name']}")
        
        try:
            response = requests.post(
                f"{BASE_URL}/policies/{policy_id}/evaluate",
                json={"input_data": test_case["data"]}
            )
            response.raise_for_status()
            result = response.json()
            
            print(f"Access allowed: {result['allowed']}")
            print(f"Reasons: {', '.join(result['reasons'])}")
            
            # Verify the result matches the expected outcome
            if result["allowed"] == test_case["expected_result"]:
                print("✅ Result matches expected outcome")
            else:
                print(f"❌ Result does not match expected outcome ({test_case['expected_result']})")
            
        except requests.RequestException as e:
            print(f"Error evaluating policy: {e}")
            # For demo purposes, show mock results
            print(f"Using mock evaluation result: {test_case['expected_result']}")
            
            # Generate mock reasons based on the test case
            if test_case["name"] == "No consent":
                print("Mock reasons: User has not provided consent")
            elif test_case["name"] == "Missing purpose":
                print("Mock reasons: Purpose is not specified")
            elif test_case["name"] == "Disallowed purpose":
                print("Mock reasons: Purpose is not allowed")
    
    # 3. Update the policy to add context-based restrictions
    print("\n3. Updating the policy to add context-based restrictions...")
    
    # Add a new rule for time-based restrictions
    updated_policy = policy_request.copy()
    updated_policy["rules"].append({
        "<=": [{"var": "context.sensitivityLevel"}, 3]
    })
    updated_policy["version"] = "1.1.0"
    updated_policy["description"] += " and sensitivity level"
    
    try:
        response = requests.put(
            f"{BASE_URL}/policies/{policy_id}",
            json=updated_policy
        )
        response.raise_for_status()
        updated = response.json()
        
        print("Policy updated successfully:")
        print(f"  New version: {updated['version']}")
        print(f"  Rules count: {len(updated['rules'])}")
        
    except requests.RequestException as e:
        print(f"Error updating policy: {e}")
        print("Using mock update result for demonstration")
        print("  New version: 1.1.0")
        print("  Rules count: 4")
    
    # 4. Evaluate with context-based restrictions
    print("\n4. Evaluating updated policy with context-based restrictions...")
    
    context_test_cases = [
        {
            "name": "Low sensitivity data",
            "data": {
                "user": {
                    "consent": True,
                    "allowed_purposes": ["marketing", "analytics"]
                },
                "purpose": "marketing",
                "context": {
                    "sensitivityLevel": 2
                }
            },
            "expected_result": True
        },
        {
            "name": "High sensitivity data",
            "data": {
                "user": {
                    "consent": True,
                    "allowed_purposes": ["marketing", "analytics"]
                },
                "purpose": "marketing",
                "context": {
                    "sensitivityLevel": 5
                }
            },
            "expected_result": False
        }
    ]
    
    for test_case in context_test_cases:
        print(f"\nScenario: {test_case['name']}")
        
        try:
            response = requests.post(
                f"{BASE_URL}/policies/{policy_id}/evaluate",
                json={"input_data": test_case["data"]}
            )
            response.raise_for_status()
            result = response.json()
            
            print(f"Access allowed: {result['allowed']}")
            print(f"Reasons: {', '.join(result['reasons'])}")
            
        except requests.RequestException as e:
            print(f"Error evaluating policy: {e}")
            print(f"Using mock evaluation result: {test_case['expected_result']}")
            
            if not test_case["expected_result"]:
                print("Mock reasons: Sensitivity level exceeds allowed threshold")
    
    # 5. Clean up - delete the policy
    print(f"\n5. Cleaning up - deleting policy {policy_id}...")
    
    try:
        response = requests.delete(f"{BASE_URL}/policies/{policy_id}")
        response.raise_for_status()
        
        print("Policy deleted successfully")
        
    except requests.RequestException as e:
        print(f"Error deleting policy: {e}")
        print("Using mock deletion result for demonstration")
        print("Policy deleted successfully (mock)")

if __name__ == "__main__":
    main() 